README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
[![Travis-CI Build
Status](https://travis-ci.org/nikita-moor/ldatuning.svg?branch=master)](https://travis-ci.org/nikita-moor/ldatuning)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/ldatuning)](https://cran.r-project.org/package=ldatuning)

This package uses some metrics to find optimal number of topics for
Latent Dirichlet allocation (LDA) models.

![](README-example-1.png)

For more complete explanation see
[vignette](http://rpubs.com/siri/ldatuning).


NEWS.md
# ldatuning 1.0.2
  * Fixed bug introduced by return_models option in FindTopicNumbers_plot(#16)

# ldatuning 1.0.1
  * Option to return list of trained models (#2)
  * Expose scoring algorithms to public namespace (#15, @titaniumtroop)
  
# ldatuning 1.0.0
  * Load balancing (#5) implemented by calculating topics in descending order
  * Parallel processing for multi-node clusters (#12, @titaniumtroop)


R/arun2010.R
# Implementation of the method, described in:
#
# Arun, R.; Suresh, V.; Veni Madhavan, C. E. & Narasimha Murthy, M. N. On
# Finding the Natural Number of Topics with Latent Dirichlet Allocation: Some
# Observations. Zaki, M. J.; Yu, J. X.; Ravindran, B. & Pudi, V. (Eds.) In
# Advances in Knowledge Discovery and Data Mining, Springer Berlin Heidelberg,
# 2010, 6118, 391-402
#
# ToDo:
# 1. make SLAM library optional (as optimization for sparse-matrix)
#    Is slam used by tm and topicmodels?
# 2. add "estimate.beta = TRUE" in control
# 3. dtm/tdm detection
# 4. It seems normalization of cm1 can be omitted

#' arun2010divergence
#'
#' symmetric Kullback-Leibler divergence
#' @keywords internal
arun2010divergence <- function(p,q) {
  entropy::KL.plugin(p,q) + entropy::KL.plugin(q,p)
}

#' arun2010
#'
#' Calculates arun2010 metric for a series of LDA models to estimate the most
#' preferable number of topics.
#'
#' @param dtm An object of class "\link[tm]{DocumentTermMatrix}" with
#'   term-frequency weighting or an object coercible to a
#'   "\link[slam]{simple_triplet_matrix}" with integer entries.
#' @param topics A list with number of topics to compare different models.
#' @param control A named list of the control parameters for estimation or an
#'   object of class "\linkS4class{LDAcontrol}".
#' @param mc.cores Integer; The number of CPU cores to processes models
#'   simultaneously (using \code{mclapply}).
#' @param verbose If false (default), supress all warnings and additional
#'   information.
#'
#' @return Data-frame with numbers of topics and corresponding values of metric
#'   (lower is better). Can be directly used by \code{\link{arun2010plot}} to
#'   draw a plot.
#'
#' @section References:
#'
#'   Arun, R., Suresh, V., Veni Madhavan, C. E. & Narasimha Murthy, M. N.
#'   (2010). On Finding the Natural Number of Topics with Latent Dirichlet
#'   Allocation: Some Observations. Zaki, M. J., Yu, J. X., Ravindran, B. &
#'   Pudi, V. (Eds.) In \emph{Advances in Knowledge Discovery and Data Mining},
#'   Springer Berlin Heidelberg, 6118, 391-402. DOI:
#'   \href{http://dx.doi.org/10.1073/10.1007/978-3-642-13657-3_43}{10.1007/978-3-642-13657-3_43}.
#'
#' @examples
#' library(topicmodels)
#' data("AssociatedPress", package="topicmodels")
#' dtm <- AssociatedPress[1:10, ]
#' arun2010(dtm, topics = 2:10, mc.cores = 2L)
#'
#' @export
arun2010 <- function(dtm, topics=seq(10, 40, by = 10), control=NULL,
                     mc.cores=2L, verbose=FALSE) {
  # check parameters
  if (length(topics[topics < 2]) != 0) {
    if (verbose) cat("warning: topics count can't to be less than 2, incorrect values was removed.\n")
    topics <- topics[topics >= 2]
  }
  # length of documents in words
  len <- slam::row_sums(dtm, dim=1)
  # fit models
  models <- parallel::mclapply(
    topics,
    FUN = function(x) topicmodels::LDA(dtm, k=x, method="Gibbs", control=control),
    mc.cores = mc.cores
  )
  # evaluate metrics
  metrics <- parallel::mclapply(
    models,
    FUN = function(model) {
      # matrix M1 topic-word
      m1 <- exp(model@beta) # rowSums(m1) == 1
      m1.svd <- svd(m1)
      cm1 <- as.matrix(m1.svd$d)
      # matrix M2 document-topic
      m2   <- model@gamma   # rowSums(m2) == 1
      cm2  <- len %*% m2    # crossprod(len, m2)
      norm <- norm(as.matrix(len), type="F")
      cm2  <- as.vector(cm2 / norm)
      return ( arun2010divergence(cm1, cm2) )
    },
    mc.cores = mc.cores
  )
  result <- data.frame(topics, as.numeric(metrics))
  names(result) <- c("topics", "arun2010")
  return(result)
}

#' arun2010plot
#'
#' Is a supplementary function to illustrate results of \code{\link{arun2010}}
#' function.
#'
#' @param results A data.frame with first column "topics" and second "arun2010".
#' @return Returns \code{\link[ggplot2]{ggplot}} object, which can be print or
#'   used for subsequent tuning by the user desire.
#'
#' @examples
#' library(topicmodels)
#' data("AssociatedPress", package="topicmodels")
#' dtm <- AssociatedPress[1:10, ]
#' metrics <- arun2010(dtm, topics = 2:10, mc.cores = 2L)
#' arun2010plot(metrics)
#'
#' @export
arun2010plot <- function(results) {
  if (!requireNamespace("ggplot2", quietly = TRUE)) {
    stop("function 'arun2010plot' requires 'ggplot2' library.")
  }
  p <- ggplot2::ggplot(data=results, ggplot2::aes_string(x="topics", y="arun2010"))
  p <- p + ggplot2::geom_line()
  p <- p + ggplot2::labs(
    title = "Number of topics (by Arun2010)",
    x = "number of topics",
    y = "symmetric Kullback-Leiber divergence"
  )
  p <- p + ggplot2::scale_x_continuous(breaks=results$topics)
  p + ggplot2::theme_bw()
}


R/griffiths2004.R
# Implementation of the method, described in:
#
# Griffiths, T. L. & Steyvers, M. Finding scientific topics Proceedings of the
# National Academy of Sciences, National Academy of Sciences, 2004, 101,
# 5228-5235
#
# Ponweiser, M. Latent Dirichlet Allocation in R. PhdThesis. Institute for
# Statistics and Mathematics, University of Economics and Business, 2012
#
#
# ToDo:
# 1. make Rmpfr library optional (looks like precision is enough)
# 2. dtm/tdm detection
# 3. It seems normalization of cm1 can be omitted
# 4. Hmisc

#' @keywords internal
griffiths2004harmonic <- function(logLikelihoods, precision=2000L) {
  # code is a little tricky, see explanation in Ponweiser2012
  # ToDo: add variant without "Rmpfr"
  llMed <- median(logLikelihoods)
  as.double( llMed - log( Rmpfr::mean( exp(
    -Rmpfr::mpfr(logLikelihoods, prec=precision) + llMed
  ))))
}

#' griffiths2004
#'
#' Calculates griffiths2004 metric for a series of LDA models to estimate the
#' most preferable number of topics.
#'
#' @inheritParams arun2010
#'
#' @return Data-frame with numbers of topics and corresponding values of metric
#'   (higher is better). Can be directly used by \code{\link{griffiths2004plot}}
#'   to draw a plot.
#'
#' @section References:
#'
#'   Griffiths, T. L. & Steyvers, M. Finding scientific topics (2004). In
#'   \emph{Proceedings of the National Academy of Sciences}, National Academy of
#'   Sciences, 101, 5228-5235. DOI:
#'   \href{http://dx.doi.org/10.1073/pnas.0307752101}{10.1073/pnas.0307752101}.
#'
#'   Ponweiser, M. Latent Dirichlet Allocation in R (2012). Diploma Thesis.
#'   Institute for Statistics and Mathematics, University of Economics and
#'   Business, Vienna. URL: \url{http://epub.wu.ac.at/id/eprint/3558}.
#'
#' @examples
#' library(topicmodels)
#' data("AssociatedPress", package="topicmodels")
#' dtm <- AssociatedPress[1:10, ]
#' griffiths2004(dtm, topics = 2:10, mc.cores = 2L)
#'
#' @export
griffiths2004 <- function(dtm, topics=seq(10, 40, by = 10), control=NULL,
                          mc.cores=2L, verbose=FALSE) {
  # check parameters
  if (length(topics[topics < 2]) != 0) {
    if (verbose) cat("warning: topics count can't to be less than 2, incorrect values was removed.\n")
    topics <- topics[topics >= 2]
  }
  if (is.null(control)) control = list()
  if (!"burnin" %in% names(control)) control$burnin <- 1000
  if (!"iter"   %in% names(control)) control$iter   <- 1000
  if (!"keep"   %in% names(control)) control$keep   <- 50
  # fit models
  models <- parallel::mclapply(
    topics,
    FUN = function(x) topicmodels::LDA(dtm, k=x, method="Gibbs", control=control),
    mc.cores = mc.cores
  )
  # log-likelihoods
  logLiks <- lapply(models, function(model) {
    model@logLiks[-c(1:(control$burnin / control$keep))]
  })
  # harmonic means
  harmony <- sapply(logLiks, function(x) griffiths2004harmonic(x))
  return(data.frame(topics, harmony))
}

#' griffiths2004plot
#'
#' Is a supplementary function to illustrate results of
#' \code{\link{griffiths2004}} function.
#'
#' @param results A data.frame with first column "topics" and second (or many)
#'   contains correspond harmonic mean values (griffiths2004 metrics).
#' @param show.max If true (default), find maximum value of metric and point it
#'   on the plot.
#' @param verbose If true, produce an additional information.
#' @return Returns \code{\link[ggplot2]{ggplot}} object, which can be print or
#'   used for subsequent tuning by the user desire.
#'
#' @examples
#' library(topicmodels)
#' data("AssociatedPress", package="topicmodels")
#' dtm <- AssociatedPress[1:10, ]
#' metrics <- griffiths2004(dtm, topics = 2:10, mc.cores = 2L)
#' griffiths2004plot(metrics)
#'
#' @export
#' @importFrom Hmisc smedian.hilow
#' @importFrom ggplot2 median_hilow
griffiths2004plot <- function(results, show.max = TRUE, verbose = FALSE) {
  # Function smedian.hilow (Hmisc) is wrapped by ggplot2, the above used
  # @importFrom works well, except for knitr (vignettes).
  if (!requireNamespace("reshape2", quietly = TRUE)) {
    stop("function 'griffiths2004plot' requires 'reshape2' package.")
  }
  p <- ggplot2::ggplot(
    reshape2::melt(results, id.vars = "topics"),
    ggplot2::aes_string(x = "topics", y = "value")
  )
  p <- p + ggplot2::geom_point()
  p <- p + ggplot2::labs(
    title = "Number of topics (by Griffiths2004)",
    x = "number of topics",
    y = "harmonic means"
  )
  p <- p + ggplot2::stat_summary(fun.data="median_hilow", geom="line", colour="black")
  p <- p + ggplot2::stat_summary(fun.data="median_hilow", geom="ribbon", alpha = 0.1)
  p <- p + ggplot2::scale_x_continuous(breaks=results$topics)

  # select the best number of topics
  if (show.max) {
    if (dim(results)[2] == 2) {
      # normal output of griffiths2004
      bestTopic <- results[ which.max(t(results[2])),"topics"]
    } else {
      # data is a result of multi-fitting
      median <- apply(results[,-1], 1, function(x) Hmisc::smedian.hilow(x))
      bestTopic <- which.max(median["Median",])
      bestTopic <- results[bestTopic,"topics"]
    }
    if (verbose) cat(sprintf("info: the recommend number of topics is %d.\n", bestTopic))
    p <- p + ggplot2::geom_vline(xintercept = bestTopic, linetype = "longdash")
  }
  p + ggplot2::theme_bw()
}


R/main.R
# Copyright (c) 2017  Nikita Murzintcev

# ToDo:
# 1. add conversion: TermDocumentMatrix > DocumentTermMatrix
#    if (is(dtm, "TermDocumentMatrix")) dtm <- t(dtm)
# 2. CaoJuan2009: check with lsa::cosine - http://stackoverflow.com/questions/2535234/find-cosine-similarity-in-r


#' FindTopicsNumber
#'
#' Calculates different metrics to estimate the most preferable number of topics
#' for LDA model.
#'
#' @param dtm An object of class "\link[tm]{DocumentTermMatrix}" with
#'   term-frequency weighting or an object coercible to a
#'   "\link[slam]{simple_triplet_matrix}" with integer entries.
#' @param topics Vector with number of topics to compare different models.
#' @param metrics String or vector of possible metrics: "Griffiths2004",
#'   "CaoJuan2009", "Arun2010", "Deveaud2014".
#' @param method The method to be used for fitting; see \link[topicmodels]{LDA}.
#' @param control A named list of the control parameters for estimation or an
#'   object of class "\linkS4class{LDAcontrol}".
#' @param mc.cores NA, integer or, cluster; the number of CPU cores to process models
#'   simultaneously. If an integer, create a cluster on the local machine. If a
#'   cluster, use but don't destroy it (allows multiple-node clusters). Defaults to
#'   NA, which triggers auto-detection of number of cores on the local machine.
#' @param return_models Whether or not to return the model objects of class
#'   "\link[topicmodels]{LDA}. Defaults to false. Setting to true requires the tibble package.
#' @param verbose If false (default), suppress all warnings and additional
#'   information.
#' @param libpath Path to R packages (use only if your R installation can't find
#'   'topicmodels' package, [issue #3](https://github.com/nikita-moor/ldatuning/issues/3).
#'   For example: "C:/Program Files/R/R-2.15.2/library" (Windows),
#'                "/home/user/R/x86_64-pc-linux-gnu-library/3.2" (Linux)
#'
#' @return Data-frame with one or more metrics.  numbers of topics and
#'   corresponding values of metric. Can be directly used by
#'   \code{\link{FindTopicsNumber_plot}} to draw a plot.
#'
#' @examples
#' \dontrun{
#'
#' library(topicmodels)
#' data("AssociatedPress", package="topicmodels")
#' dtm <- AssociatedPress[1:10, ]
#' FindTopicsNumber(dtm, topics = 2:10, metrics = "Arun2010", mc.cores = 1L)
#' }
#'
#' @export
#' @import topicmodels
FindTopicsNumber <- function(dtm, topics = seq(10, 40, by = 10),
                             metrics = "Griffiths2004",
                             method = "Gibbs", control = list(),
                             mc.cores = NA, return_models = FALSE,
                             verbose = FALSE, libpath = NULL) {
  # check parameters
  if (length(topics[topics < 2]) != 0) {
    if (verbose) cat("warning: topics count can't to be less than 2, incorrect values was removed.\n")
    topics <- topics[topics >= 2]
  }
  topics <- sort(topics, decreasing = TRUE)

  if ("Griffiths2004" %in% metrics) {
    if (method == "VEM") {
      # memory allocation error
      if (verbose) cat("'Griffiths2004' is incompatible with 'VEM' method, excluded.\n")
      metrics <- setdiff(metrics, "Griffiths2004")
    } else {
      # save log-likelihood when generating model
      if (!"keep" %in% names(control)) control <- c(control, keep = 50)
    }
  }

  # fit models
  if (verbose) cat("fit models...")

  # Parallel setup
  if (any(class(mc.cores) == "cluster")) {
    cl <- mc.cores
  } else if (isTRUE(class(mc.cores) == "integer")) {
    cl <- parallel::makeCluster(mc.cores)
  } else {
    cl <- parallel::makeCluster(parallel::detectCores())
  }
  parallel::setDefaultCluster(cl)
  parallel::clusterExport(varlist = c("dtm", "method", "control"),
                          envir = environment())
  models <- parallel::parLapply(cl = cl, X = topics, fun = function(x) {
  if (is.null(libpath) == FALSE) { .libPaths(libpath) }
    topicmodels::LDA(dtm, k = x, method = method, control = control)
  })
  if (! any(class(mc.cores) == "cluster")) {
    parallel::stopCluster(cl)
  }
  if (verbose) cat(" done.\n")

  # calculate metrics
  if (verbose) cat("calculate metrics:\n")

  if (return_models &
      requireNamespace("tibble", quietly = TRUE)
  ) {
    result <- cbind(topics, tibble::enframe(models, value = "LDA_model"))
    result$name <- NULL
  } else {
    if (return_models) {
      message("The tibble package is required for returning models. Returning results only.")
    }
    result <- data.frame(topics)
  }
  for(m in metrics) {
    if (verbose) cat(sprintf("  %s...", m))
    if (! m %in% c("Griffiths2004", "CaoJuan2009", "Arun2010", "Deveaud2014")) {
      cat(" unknown!\n")
    } else {
      result[m] <- switch(m,
        "Griffiths2004" = Griffiths2004(models, control),
        "CaoJuan2009"   = CaoJuan2009(models),
        "Arun2010"      = Arun2010(models, dtm),
        "Deveaud2014"   = Deveaud2014(models),
        NaN
      )
      if (verbose) cat(" done.\n")
    }
  }

  return(result)
}

#' Griffiths2004
#'
#' Implement scoring algorithm. In order to use this algorithm, the LDA model MUST
#' be generated using the keep control parameter >0 (defaults to 50) so that the
#' logLiks vector is retained.
#' @param models An object of class "\link[topicmodels]{LDA}
#' @param control A named list of the control parameters for estimation or an
#'   object of class "\linkS4class{LDAcontrol}".
#' @return A scalar LDA model score
#'
#' @export
#'
Griffiths2004 <- function(models, control) {
  # log-likelihoods (remove first burnin stage)
  burnin  <- ifelse("burnin" %in% names(control), control$burnin, 0)

  logLiks <- lapply(models, function(model) {
    # Check to make sure logLiks were kept; if not, value is NaN
    if (length(model@logLiks) == 0) {
      message("No logLiks were kept, which is required to use this scoring algorithm. Please regenerate the model using the keep control parameter set to a reasonable value (default = 50).")
      NaN
    } else {
      utils::tail(model@logLiks, n = length(model@logLiks) - burnin/control$keep)
      # model@logLiks[-(1 : (control$burnin/control$keep))]
    }
  })

  # harmonic means for every model
  metrics <- sapply(logLiks, function(x) {
    # code is a little tricky, see explanation in [Ponweiser2012 p. 36]
    # ToDo: add variant without "Rmpfr"
    llMed <- stats::median(x)
    metric <- as.double(
      llMed - log( Rmpfr::mean( exp( -Rmpfr::mpfr(x, prec=2000L) + llMed )))
    )
    return(metric)
  })
  return(metrics)
}

#' CaoJuan2009
#'
#' Implement scoring algorithm
#' @param models An object of class "\link[topicmodels]{LDA}
#' @return A scalar LDA model score
#'
#' @export
#'
CaoJuan2009 <- function(models) {
  metrics <- sapply(models, function(model) {
    # topic-word matrix
    m1 <- exp(model@beta)
    # pair-wise cosine distance
    pairs <- utils::combn(nrow(m1), 2)
    cos.dist <- apply(pairs, 2, function(pair) {
      x <- m1[pair[1], ]
      y <- m1[pair[2], ]
      # dist <- lsa::cosine(x, y)
      dist <- crossprod(x, y) / sqrt(crossprod(x) * crossprod(y))
      return(dist)
    })
    # metric
    metric <- sum(cos.dist) / (model@k*(model@k-1)/2)
    return(metric)
  })
  return(metrics)
}

#' Arun2010
#'
#' Implement scoring algorithm
#' @param models An object of class "\link[topicmodels]{LDA}
#' @param dtm An object of class "\link[tm]{DocumentTermMatrix}" with
#'   term-frequency weighting or an object coercible to a
#'   "\link[slam]{simple_triplet_matrix}" with integer entries.
#' @return A scalar LDA model score
#'
#' @export
#'
Arun2010 <- function(models, dtm) {
  # length of documents (count of words)
  len <- slam::row_sums(dtm)
  # evaluate metrics
  metrics <- sapply(models, FUN = function(model) {
    # matrix M1 topic-word
    m1 <- exp(model@beta) # rowSums(m1) == 1
    m1.svd <- svd(m1)
    cm1 <- as.matrix(m1.svd$d)
    # matrix M2 document-topic
    m2   <- model@gamma   # rowSums(m2) == 1
    cm2  <- len %*% m2    # crossprod(len, m2)
    norm <- norm(as.matrix(len), type="m")
    cm2  <- as.vector(cm2 / norm)
    # symmetric Kullback-Leibler divergence
    divergence <- sum(cm1*log(cm1/cm2)) + sum(cm2*log(cm2/cm1))
    return ( divergence )
  })
  return(metrics)
}

#' Deveaud2014
#'
#' Implement scoring algorithm
#' @param models An object of class "\link[topicmodels]{LDA}
#' @return A scalar LDA model score
#'
#' @export
#'
Deveaud2014 <- function(models) {
  metrics <- sapply(models, function(model) {
    ### original version
    # topic-word matrix
    m1 <- exp(model@beta)
    # prevent NaN
    if (any(m1 == 0)) { m1 <- m1 + .Machine$double.xmin }
    # pair-wise Jensen-Shannon divergence
    pairs  <- utils::combn(nrow(m1), 2)
    jsd <- apply(pairs, 2, function(pair) {
      x <- m1[pair[1], ]
      y <- m1[pair[2], ]
      ### standard Jensen-Shannon divergence
      # m <- (x + y) / 2
      # jsd <- 0.5 * sum(x*log(x/m)) + 0.5 * sum(y*log(y/m))
      ### divergence by Deveaud2014
      jsd <- 0.5 * sum(x*log(x/y)) + 0.5 * sum(y*log(y/x))
      return(jsd)
    })

#     ### optimized version
#     m1   <- model@beta
#     m1.e <- exp(model@beta)
#     pairs  <- utils::combn(nrow(m1), 2)
#     jsd <- apply(pairs, 2, function(pair) {
#       x   <- m1[pair[1], ]
#       y   <- m1[pair[2], ]
#       x.e <- m1.e[pair[1], ]
#       y.e <- m1.e[pair[2], ]
#       jsd <- ( sum(x.e*(x-y)) + sum(y.e*(y-x)) ) / 2
#       return(jsd)
#     })

    # metric
    metric <- sum(jsd) / (model@k*(model@k-1))
    return(metric)
  })
  return(metrics)
}


#' FindTopicsNumber_plot
#'
#' Support function to analyze optimal topic number. Use output of the
#' \code{\link{FindTopicsNumber}} function.
#'
#' @param values Data-frame with first column named `topics` and other columns
#'   are values of metrics.
#'
#' @examples
#' \dontrun{
#'
#' library(topicmodels)
#' data("AssociatedPress", package="topicmodels")
#' dtm <- AssociatedPress[1:10, ]
#' optimal.topics <- FindTopicsNumber(dtm, topics = 2:10,
#'   metrics = c("Arun2010", "CaoJuan2009", "Griffiths2004")
#' )
#' FindTopicsNumber_plot(optimal.topics)
#' }
#'
#' @export
#' @import ggplot2
FindTopicsNumber_plot <- function(values) {
  # Drop models if present, as they won't rescale
  if ("LDA_model" %in% names(values)) {
    values <- values[!names(values) %in% c("LDA_model")]
  }
  # normalize to [0,1]
  columns <- base::subset(values, select = 2:ncol(values))
  values <- base::data.frame(
    values["topics"],
    base::apply(columns, 2, function(column) {
      scales::rescale(column, to = c(0, 1), from = range(column))
    })
  )

  # melt
  values <- reshape2::melt(values, id.vars = "topics", na.rm = TRUE)

  # separate max-arg & min-arg metrics
  values$group <- values$variable %in% c("Griffiths2004", "Deveaud2014")
  values$group <- base::factor(
    values$group,
    levels = c(FALSE, TRUE),
    labels = c("minimize", "maximize")
  )

  # standart plot
  p <- ggplot(values, aes_string(x = "topics", y = "value", group = "variable"))
  p <- p + geom_line()
  p <- p + geom_point(aes_string(shape = "variable"), size = 3)
  p <- p + guides(size = FALSE, shape = guide_legend(title = "metrics:"))
  p <- p + scale_x_continuous(breaks = values$topics)
  p <- p + labs(x = "number of topics", y = NULL)

  # separate in two parts
  p <- p + facet_grid(group ~ .)

  # style
  # p <- p + theme_bw(base_size = 14, base_family = "") %+replace% theme(
  p <- p + theme_bw() %+replace% theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    panel.grid.major.x = element_line(colour = "grey70"),
    panel.grid.minor.x = element_blank(),
    legend.key = element_blank(),
    strip.text.y = element_text(angle = 90)
  )

  # move strip block to left side
  g <- ggplotGrob(p)
  g$layout[g$layout$name == "strip-right", c("l", "r")] <- 3
  grid::grid.newpage()
  grid::grid.draw(g)

  # return(p)
}
